Added copy-to-clipboard functionality

lukasz-migas opened the pull request:
- added copy screenshot of viewer/window to the clipboard
- fixed unrelated typo

# Description
This PR adds the funcionality to copy screenshot of the viewer (or window) to the user's clipboard. This is accessible via `Menu -> File -> Copy Screenshot to Clipboard` or via `viewer.window.clipboard()` to capture screenshot of the entire window or `viewer.window.qt_viewer.clipboard()` to only capture the canvas.

## Type of change
- [x] New feature (non-breaking change which adds functionality)

# How has this been tested?
Added extra test to make sure an iamge is copied into the user's clipboard.

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If I included new strings, I have used `trans.` to make them localizable.

lukasz-migas committed:
Added copy-to-clipboard functionality

- added copy screenshot of viewer/window to the clipboard
- fixed unrelated typo

lukasz-migas committed:
Update test_qt_viewer.py

lukasz-migas committed:
Merge branch 'master' into viewer-clipboard

lukasz-migas committed:
Update test_qt_viewer.py

goanpeca commented on napari/_qt/qt_main_window.py:
```diff
@@ -488,6 +488,14 @@ def _add_file_menu(self):
         screenshot_wv.triggered.connect(self._screenshot_dialog)
 
+        clipboard = QAction(
+            trans._('Copy Screenshot to Clipboard...'), self._qt_window
```
The `...` are used to indicate that a new dialog will be opened, so we can remove it :)

lukasz-migas replied:
Thanks! Indeed it doesn't make sense for the `...` to be there.

goanpeca commented on napari/_qt/qt_main_window.py:
```diff
@@ -488,6 +488,14 @@ def _add_file_menu(self):
+        clipboard.setStatusTip(
+            trans._('Copy screenshot of current display to the clipboard')
+        )
+        clipboard.triggered.connect(self.qt_viewer.clipboard)
```
This is triggering the screenshot for the viewer content and not the entire window, `self.qt_viewer.clipboard` vs `self.clipboard`

Just want to make sure, so that the action text is clear `Copy Screenshot to Clipboard` sounds like it would run `self.clipboard`

lukasz-migas replied:
True, it does call the `qt_viewer.clipboard` method rather than the `clipboard` method. It's following the same logic as the `Screenshot` menu options. Perhaps it would make sense to change the text to `Copy Screenshot of the Canvas to Clipboard` although that is quite long menu entry.

goanpeca replied:
Indeed. If it is following the same convention as a preexisting action then we are fine :) and no change is needed.

goanpeca reviewed (changes requested):
Thanks for working on this @lukasz-migas 😃 !

Left a couple of questions and a minor change.

I was wondering if we could make the fact that a screenshot was taken by flashing either, the main window, or just the viewer (depending on which clipboard method is called)

That would really help in making things apparent.

You could create a helper method like the one in here
https://github.com/napari/napari/pull/2284/files#diff-91ff0176aad5008b16e25075ab1de158e640bc75a256a179f58a46670fc4970eR245

So that you could pass any widget to it and reuse it (`napari/_qt/utils`) to flash/blink a widget.

Thanks!

lukasz-migas committed:
Merge branch 'master' into viewer-clipboard

lukasz-migas committed:
Added flash animation to highlight screenshot region

* added `add_flash_animation` function to _qt/utils that permits adding quick flash to Qt widgets
* minor improvements to text.
* added example to `examples` to illustrate the clipboard functionality.

lukasz-migas committed:
Fixed flash issue

* fixed the issue where flash animation could not be applied to the `native` widget.

goanpeca commented on napari/_qt/_tests/test_qt_viewer.py:
```diff
@@ -303,6 +304,24 @@ def test_points_layer_display_correct_slice_on_scale(make_napari_viewer):
     np.testing.assert_equal(indices, [0])
 
 
+def test_qt_viewer_clipboard(make_napari_viewer):
```
```
 ==================================== ERRORS ====================================
  ________________ ERROR at teardown of test_qt_viewer_clipboard _________________
 ```

It seems this tests has some leaked widgets. Suggestions @tlambert03? Thanks :)

goanpeca reviewed (approved):
🚀

sofroniewn commented on examples/clipboard.py:
```diff
@@ -0,0 +1,24 @@
+
+layer_buttons = viewer.window.qt_viewer.layerButtons
+
+# Add button to copy image of the canvas to clipboard.
```
I'm not so sure about this example - i find the buttons with the warning symbol a little confusing to connect to the copy to clip board functionality. Maybe we just do without it as the functionality is in a dropdown menu?

lukasz-migas replied:
Thanks for the feedback - perhaps it wasn't the best example. I've modified it to add include buttons with text in a side dock instead.

sofroniewn commented on napari/_qt/qt_main_window.py:
```diff
@@ -529,6 +548,8 @@ def _add_file_menu(self):
         self.file_menu.addAction(save_all_layers)
         self.file_menu.addAction(screenshot)
         self.file_menu.addAction(screenshot_wv)
+        self.file_menu.addAction(clipboard)
+        self.file_menu.addAction(clipboard_wv)
```
I'm more used to thinking of copy actions as being present in an edit menu, instead of a filemenu, but i get the screenshot in general is in the file menu. What does @goanpeca think here?

goanpeca replied:
Agreed but it may be something for a different PR, since we would need to move other issues and make sure they are disabled if canvas is empty

sofroniewn replied:
Interesting, I hadn't realized that, but that's fine, can move later

goanpeca replied:
Will open an issue to track this.

sofroniewn commented on napari/_qt/utils.py:
```diff
@@ -252,6 +253,47 @@ def combine_widgets(
     )
 
 
+def add_flash_animation(widget: QWidget):
```
Love the functionality, but can you make `duration` and `color` parameters to the `add_flash_animation`. They can have sensible defaults, but I imagine that one day we might adjust these things/ it would be nice if they were easy to change.

I also find the flash quite intense right now. I feel a lot better with duration `300ms` and the color being `QColor(128, 128, 128, 255)`. It's still very noticeable but much easier on my eyes!

goanpeca replied:
Sounds good!

lukasz-migas replied:
Sure, I've modifed the function to take in duration and color parameters which should make it more general (e.g. tutorial widget in the future?)
Indeed, the flash to white was very intense so its been changed to (128, 128, 128, 128) - it's still visible but not offensive.

sofroniewn commented on napari/_qt/qt_viewer.py:
```diff
@@ -571,6 +571,22 @@ def screenshot(self, path=None):
+        """
+        from .utils import add_flash_animation
+
+        img = self.canvas.native.grabFramebuffer()
```
can we maybe put this line inside a private `_screenshot` method that we then use inside our public `screenshot(self, path=None)` and here too (i.e. so as not to duplicate `img = self.canvas.native.grabFramebuffer()`). I'd like to keep instances of going into `.native` as few as possible in the code.

Also I'm wondering if the flash shouldn't be inside that method too, could be a parameter

```
def _screenshot(self, flash=True):
...
    returns img
```

tlambert03 replied:
This is great (as usual) @lukasz-migas!  @sofroniewn's comment here was my only comment/thought as well: namely, that `clipboard()` and `screenshot()` methods should be using the same function behind the scenes to actually get the image.  Otherwise, great stuff!

lukasz-migas replied:
Thanks for the feedback - I've refactored the `clipboard()` and `screenshot()` functions to use the same private method behind the scenes. I've also added `flash=True` to the `screenshot()` method to highlight when screenshot has taken place.

sofroniewn reviewed (commented):
Looking great, as always @lukasz-migas really nice work for another lovely contribution. Apologies for my delay in reviewing. I've left some fairly minor comments. The flash idea is great, but I'd like to make it a little softer on the eyes. I'm also wondering if it should play every time you take a screenshot (even if saving to disk). I know there a dialog has opened up from file menu, but won't be case if using the API directly and so it could be nice to see too.

lukasz-migas committed:
Refactored screenshot/clipboard functionality

- moved screenshot functionality to own private method
- added flash functionality to screenshot
- flash animation was changed to use light grey instead of white

lukasz-migas committed:
Merge branch 'napari:master' into viewer-clipboard

lukasz-migas committed:
Minor bug fixes

* added the missing `flash` argument to the `screenshot` method in the viewer
* disabled flash in notebook display
* disabled flash in tests since it would require a lot of `qtbot.wait`

sofroniewn reviewed (approved):
This looks great @lukasz-migas! This is the current behavior of `examples/clipboard.py`, which I think it very nice. Thanks for making the toy dockwidget too - great way to demo the functionality

https://user-images.githubusercontent.com/6531703/119767368-89f14280-be6b-11eb-8f10-0d9740a57786.mov

tlambert03 reviewed (approved):
LGTM!

tlambert03 committed:
Merge branch 'master' into viewer-clipboard
